File: AtCoder/Typical90/002/002.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

int main(){
    int n;
    cin >> n;
    vector<string> ans;
    if(n%2 == 1) return 0;
    for(int bit=0; bit<(1<<n); bit++){
        int cnt = 0;
        bool is_correct = true;
        for(int i=0; i<n; i++){
            if(bit & (1<<i)){
                cnt++;
            }
            else{
                cnt--;
            }
            if(cnt < 0) is_correct = false;
        }
        if(is_correct && cnt == 0){
            string s;
            for(int i=0; i<n; i++){
                if(bit & (1<<i)){
                    s += '(';
                }
                else{
                    s += ')';
                }
            }
            ans.push_back(s);
        }
    }
    sort(ans.begin(), ans.end());
    for(auto s: ans){
        cout << s << endl;  
    }
}

File: AtCoder/Typical90/002/002.rb
n = gets.chomp.to_i
return if n.odd?
ans = Array.new
[0, 1].repeated_permutation(n).each do |perm|
    cnt = 0
    is_correct = true
    perm.to_a.each do |val|
        cnt += (val == 0 ? 1 : -1)
        is_correct = false if cnt < 0 
    end
    if is_correct and cnt == 0
        s = String.new
        perm.each do |val|
        s += (val == 0 ? '(' : ')')
        end
        ans << s
    end
end
ans.sort.each do |s|
    puts s
end